test/parameterized_tester.c

```diff
@@ -0,0 +1,194 @@
+/* parameterized_tester.c - Test deflate() and inflate() on a chunk of the
+ * input data file, combining many deflate() settings to improve testing coverage.
+ */
+#include "zbuild.h"
+#ifdef ZLIB_COMPAT
+#  include "zlib.h"
+#else
+#  include "zlib-ng.h"
+#endif
+
+#include <stdio.h>
+
+/*
+ * WITH_EXTRA_TESTS results in 9450 permutations per input file.
+ * Normal mode results in 540.
+ *
+ * WITH_EXTRA_TESTS also tests a bigger chunk of the files.
+ */
+
+#ifdef WITH_EXTRA_TESTS
+const int valid_wbits[] = {
+    -9, -10, -11, -12, -13, -14, -15,
+     9,  10,  11,  12,  13,  14,  15,
+#  ifdef WITH_GZFILEOP
+    25,  26,  27,  28,  29,  30,  31,
+#  endif
+};
+
+const int valid_compression_levels[] = {
+    0, 1, 2, 3, 4, 5, 6, 7, 8, 9
+};
+
+const int valid_mem_levels[] = {
+    1, 2, 3, 4, 5, 6, 7, 8, 9,
+};
+
+#define MAX_FILESIZE_KB 4096
```

Comment:
Do we need this?

Why not just allocate dynamically enough memory for the entire file?

We already open it using fopen so we can fseek and ftell to get the size and then allocate memory.

Reply:
Running 9450 iterations on 50+MB files would be extremely timeconsuming.

Reply:
Time consuming or not, I would expect at least one iteration to fail without any code changes... That would warrant adding complex testing that essentially limits the corner cases we have tested before. Personally I'm not confident enough that limiting file size to just 4 MB (4096 kB) is enough.

Reply:
We might roll back to using the original testing method instead then, and let this tool be something developers can run manually, like it was in #1448.
Because doing complete testing like this would just not be possible through github actions, the jobs would likely take more than the max 6hr job limit, and we would for sure quickly tie up our 10 runners and block all CI progress.